internal/bsr/bsr.go

```diff
@@ -237,6 +238,140 @@ func OpenSession(ctx context.Context, sessionRecordingId string, f storage.FS, k
 	return session, nil
 }
 
+// Validation provides the results from validating a bsr.
+type Validation struct {
+	SessionRecordingId         string
+	Valid                      bool
+	SessionRecordingValidation *ContainerValidation
+}
+
+// ContainerValidation contains the results from validating a container in a bsr.
+type ContainerValidation struct {
+	Name                    string
+	ContainerType           ContainerType
+	Error                   error
+	FileChecksumValidations ContainerChecksumValidation
+	SubContainers           []*ContainerValidation
+}
+
+// Validate retrieves a BSR from storage using the sessionRecordingId and validates the BSR.
+// All files and sub container files will be verified by comparing it against checksums for
+// each file in SHA256SUM file.
+//
+// Validation will continue even if there's an error encountered during validation.
+// The validation error will be added to the ContainerValidation struct "Error" field for that container.
+func Validate(ctx context.Context, sessionRecordingId string, f storage.FS, keyUnwrapFn kms.KeyUnwrapCallbackFunc) (*Validation, error) {
+	const op = "bsr.Validate"
+
+	switch {
+	case sessionRecordingId == "":
+		return nil, fmt.Errorf("%s: missing session recording id: %w", op, ErrInvalidParameter)
+	case f == nil:
+		return nil, fmt.Errorf("%s: missing storage: %w", op, ErrInvalidParameter)
+	case keyUnwrapFn == nil:
+		return nil, fmt.Errorf("%s: missing key unwrap function: %w", op, ErrInvalidParameter)
+	}
+
+	session, err := OpenSession(ctx, sessionRecordingId, f, keyUnwrapFn)
+	if err != nil {
+		return nil, fmt.Errorf("%s: failed to retrieve session for %s: %w", op, sessionRecordingId, err)
+	}
+
+	validation := &Validation{
+		SessionRecordingId: sessionRecordingId,
+		Valid:              true,
+	}
+
+	sessionContainerValidation := validateContainer(ctx, validation, SessionContainer, session.container, session.Meta.Id)
+	validation.SessionRecordingValidation = sessionContainerValidation
+	if err != nil {
+		validation.Valid = false
+		return validation, fmt.Errorf("%s: failed to validate session for %s: %w", op, sessionRecordingId, err)
+	}
+
+	// Validate all connections under session
+	for connId := range session.Meta.connections {
+		// Get connection id
+		connKey := connId
+		lastDotIndex := strings.LastIndex(connId, ".connection")
+		if lastDotIndex == -1 {
+			validation.Valid = false
+			return validation, fmt.Errorf("%s: malformed BSR for: %s", op, connId)
+		}
+		connKey = connId[:lastDotIndex]
+
+		// Open current connection
+		connection, err := session.OpenConnection(ctx, connKey)
```

Comment:
Likewise we should be calling `connection.Close(ctx)` when we are done with the connection. You might want to refactor the contents of this for loop into a function to make it easier to use `defer` so the connection is closed before iterating to the next connection.